Rust/src/tree/tree.rs

```diff
@@ -198,22 +202,63 @@ impl<T> Tree<T>
     ///
     /// // check that we recover the only node in the tree
     /// let query = vec![0.1, 0.9];
-    /// let nodes: Vec<&Node<f32>> = tree.traverse(&query).collect();
+    /// let nodes: Vec<&Node<f32>> = tree.iter(&query).collect();
     /// assert_eq!(nodes.len(), 1);
     ///
     /// // after adding a second point the traversal should contain the
     /// // point closest to the query in the L1 norm
     /// tree.add_point(vec![-1.0, -2.0]);
-    /// let nodes: Vec<&Node<f32>> = tree.traverse(&query).collect();
+    /// let nodes: Vec<&Node<f32>> = tree.iter(&query).collect();
     /// assert_eq!(nodes.len(), 2);
     ///
     /// // a traversal implements iter, so we can use it in a loop
-    /// for node in tree.traverse(&query) {
+    /// for node in tree.iter(&query) {
     ///     println!("mass = {}", node.mass());
     /// }
     /// ```
-    pub fn traverse<'a>(&'a self, point: &'a Vec<T>) -> NodeTraverser<'a, T> {
-        NodeTraverser::new(self, point)
+    pub fn iter<'a>(&'a self, point: &'a Vec<T>) -> NodeIterator<'a, T> {
+        NodeIterator::new(self, point)
+    }
+
+    /// Apply a visitor to a tree traversal with a query point.
+    ///
+    /// Given a query point and a visitor, we apply the visitor to the nodes
+    /// of the tree starting at the leaf nearest to the query and traversing up
```

Comment:
A little bit confused. Where is the logic to start at the nearest leaf? This appears to start at the root.

Reply:
Great question. If you look at `traverse_helper()` you'll see the recursion step:

```rust
self.traverse_helper(point, visitor, next_node_key, depth + One::one());
visitor.accept(&node, depth);
```

This means that we recurse down to the nearest leaf node and execute `accept_leaf()` first before following the stack back up to execute `accept()` at each of the intermediate internal nodes.

It's different than `Tree::iter()`, I know. I kept `iter()` mostly because it helps with some tests and I couldn't figure out a *good way* to return a node iteration in the same order as `traverse()`.

Reply:
Ah I see makes sense.

Reply:
Okay, I think I am missing something. Why cant the scoring start from the root and compute as it goes down to the nearest leaf in order to avoid recursion?

Reply:
Really good question. I was writing up a response but realized that I it was (a) getting really long and should be turned into a manuscript of some kind and (b) not all that convincing that one direction is better than another. Still, a good exercise to carry out.

@sudiptoguha Do you have a one-sentence argument for this?

Reply:
Probably not a one-sentence :) The discussion is perhaps top-down versus bottom up? In general bottom up is more powerful and more expressive. One example is that in the current Java version, the bounding boxes are recreated on the fly and once a point is inside a bounding box then there is no change to score/density/most of envisioned visitor applications (forecasting is a different discussion altogether). If you start from the top, you need the bounding box at the top -- continuing the argument, you need bounding boxes everywhere. If you go bottom up -- you only need to only ever build the bounding boxes you need. For N leaves, the number of possible internal bounding boxes is N-1 and the bounding boxes are 2D size (max, min for each of the D dimensions). So the bounding boxes themselves are almost 2ND in size -- note the amount of basic information is ND (for the points, ignoring the shingling). So there you have 66% saving (ND versus ND + 2ND) savings in space if you go bottom up and build boxes on the fly. And you can adjust the space/throughput tradeoff by storing some of the boxes (as the Java version does).

Reply:
The anomaly score algorithm is defined recursively at level n by:
```
(1 - p) score_n + p score_{n + 1}
```
where p is the probably that the query point (i.e., the point being scored) would be separated from the bounding box at level n on the query path by a random cut. The original implementation of RCFs didn't have a generic traversal function, it just had a recursive anomaly score function that mirrored the score computation. We kept the leaf-to-root logic of the anomaly score method when we refactored the library to use the visitor pattern. So the answer is partly historical, but algorithms that we added to the library later (like density estimation and imputation) followed the same pattern of computing a value at the leaf node first and then updating that value by traversing back to the root. That said, it's not a requirement that the implementation of the traversal function be recursive, and there's no reason we couldn't also define a traversal method that visits nodes in root to leaf order.

Reply:
I would also add that so far we haven't seen much downside in the recursive implementation, since the recursion depth will be equal to the tree depth, which is on the order of log-sample-size. I did one test a long time ago (when the library looked a lot different than it does now) where I rewrote the traversal method to be iterative instead of recursive, but it was actually slower than the recursive version. So from my perspective I'm open to an iterative implementation, changing this part of the library hasn't been a high priority.

Reply:
@sudiptoguha Thanks for the response and for bringing up the dynamic bounding boxes as an example. Very clear reason to use the bottom up approach.

@jotok Glad that you experience with the scoring methods sort of validates that (at least specific to *anomaly* scoring) traversal direction is somewhat of a choice rather than a requirement. Of course this doesn't apply to some of the other scoring algorithms. Interesting bit about the performance difference.